mrmustard/utils/misc_tools.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains miscellaneous functions which are used accross the code base of MM. 
+They can be used as is in any of the classes.
+"""
+
+all = ["duck_type_checker", "general_factory"]
+
+
+def duck_type_checker(obj_a: object, obj_b: object) -> bool:
+    r"""*If it walks like a duck and it quacks like a duck, then it must be a duck.*
+
+    This function performs type-checking from a duck-type perspective, testing whether both given
+    objects share the same set of attributes. If they do, no matter their *type* -as obtained by
+    a call to ``isinstance``-, they will be considered to be of the same duck-type.
+
+    Note that we here look at the symmetric difference, hence the exact question we ask is better
+    phrased as "*Are both objects of the same type?*" than as "*Is object a of the same type as
+    object b?*".
+
+    Note also that this checks for the exact match between types -via attributes-, this system is
+    **not** meant to work with subclasses since there is a high probability that they might have
+    different attributes.
+
+    Args:
+        obj_a:  one of the objects to be checked
+        obj_b:  the other object to be checked
+
+    Returns:
+        True if both objects have exactly the same attributes, False otherwise.
```

Comment:
missing return type
```suggestion
    Returns:
        bool: True if both objects have exactly the same attributes, False otherwise.
```

Reply:
Ok so this brings up a wider question for me.
Whether we include the types in the doc or not.
My understanding was that we don't, to avoid redundancy of information. So while before I used to have something like:

def foo(bar:int) -> bool:
    """ Loremipsum.
    Args:
        bar (int): in dolores
    Returns:
        (bool) sint lament"""

I had changed to:
def foo(bar:int) -> bool:
    """ Loremipsum.
    Args:
        bar: in dolores
    Returns:
        sint lament"""
After reading the pennylane and a short chat with you because the way I understood it it was more valuable to avoid repeating info. So just want to make sure I understand things correctly so I can update my other code. What's our MM convention for typing in docstring? I'm happy with anything, just want to get it right ^^'.

Reply:
from the [pennylane standard](https://docs.pennylane.ai/en/stable/development/guide/documentation.html#functions-and-methods) it is customary to add the types to the docstrings (even if type hints are present in the function)

```python
def func(arg1, arg2, arg3="default_value", **kwargs):
    """Single sentence that summarizes the function.

    Multi-line description of the function (optional, only if necessary).

    Args:
        arg1 (type): Description.
            Continuation line is indented if needed.
        arg2 (type): description
        arg3 (str): Description. Do not provide the default
            value, as this is already present in the signature.

    Keyword Args:
        kwarg1 (type): Description. Provide the default value if applicable.

    Returns:
        type: Description.
        Continuation line is not indented.

    Raises:
        ExceptionType: description

    .. seealso:: :func:`~.relevant_func`, :class:`~.RelevantClass` (optional)

    **Example**

    Minimal example with 1 or 2 code blocks (required).

    .. details::
        :title: Usage Details

        More complicated use cases, options, and larger code blocks (optional).

    **Related tutorials**

    Links to any relevant PennyLane demos (optional).
    """
```

Reply:
allright!

Reply:
I've added the types hints, except for args/kwargs because I wasn't sure what's the best typing there.